First pass at protocol documentation

bostik opened the pull request:
The statsd protocol variant used in bucky3 is essentially a subset of
the DogStatsD protocol.

Add documentation on how the protocol looks like in various use cases.
We want to encourage others to write client libraries, because statsd
users really shouldn't need to know about the on-the-wire format.

Essentially - this is not meant as the final version, it's mostly a collection of loose ends to describe how the messages are constructed and what bucky3 does with them.

Commit:
First pass at protocol documentation

The statsd protocol variant used in bucky3 is essentially a subset of
the DogStatsD protocol.

Add documentation on how the protocol looks like in various use cases.
We want to encourage others to write client libraries, because statsd
users really shouldn't need to know about the on-the-wire format.

Signed-off-by: Mika Boström <bostik@iki.fi>

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+    <measurement name>:1|c|#<tag name>=<tag value>[,<tag name>=<tag value[,...]]<newline>
+
+* `measurement name` is an arbitrary string
+* `value` is the recorded measurement, if type is other than `c`
```
Please mention that bucky3 takes floats for all types of values (some flavors of statsd take ints)

bostik replied:
Will do.

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+
+### Message Format
+
+The StatsD messages are sent as UDP datagrams with ASCII encoded
```
Let's emphasize ASCII with bold font face or something like this. This detail is actually very important.

bostik replied:
Right. I'll put in an emphasis and some careful rewording.

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+
+    <measurement name>:<value>|<type>|#<tag name>=<tag value>[,<tag name>=<tag value>[,...]]<newline>
+
+    <measurement name>:1|c|#<tag name>=<tag value>[,<tag name>=<tag value[,...]]<newline>
```
What's a difference between these two line? (the latter looks like a specific case of the former)
How about the rate?
Also, it may be worth mentioning that bucky3 can take both `=` and `:` as a k=v (k:v) separator.

bostik replied:
I originally wanted to underscore that for all normal use cases `counter` is a special case due to the fixed increment. Will drop it.

As for rate, somehow I missed that. This section clearly needs some hammering. The same with the tag formatting: the regex does indeed capture the separator with `[:=]`. Stay tuned..

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+* `type` is one of the following:
+  * `g` - gauge; an already measured value
+  * `ms` - timer (milliseconds); duration of an operation
+  * `h` - histogram; distribution of (often discrete) values
```
`h` is an alias for `ms`, worth a mention?

bostik replied:
I think we can mention it, yes.

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+  * `s` - set; collection of distinct keys
+  * `c` - counter; an automatically increasing count of occurrences
+* `tag name` is an arbitrary label or tag for the value
+* `tag value` is an arbitrary categorisation
```
Names and values impose some limits (lowest common denominator of Prometheus, InfluxDB and DataDog implementations) - but it is a big topic and could be covered in next commits.

bostik replied:
Aye, let's tackle that separately.

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+code has been served. So, just before returning the service response,
+the service could send the following message to bucky3:
+
+  - `status:200|h|#service=login,team=myteam,route=/user/login\n`
```
Looks like a use case for `s` rather than `h` here.

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+overriding the counter makes sense. Even then, it is recommended that
+you use a gauge instead. Flooding bucky3 with extremely rapid
+counter-increase messages has the potential to degrade monitoring
+performance.)
```
The rate option in protocol helps in such cases, too. Counter is only type of aggregate where the use of rate actually makes clear sense.

jsiembida commented on PROTOCOL.md:
```diff
@@ -0,0 +1,127 @@
+they could send the following two messages:
+
+- `status:200|h|#service=ourstream,team=otherteam,action=something\n`
+- `duration:2.9|ms|#service=ourstream,team=otherteam,action=something\n`
```
Another option is to integrate status into duration as in `duration:1|ms|#action=something,result=200` - that works ok when the set of all `result` values is small (like in HTTP). For big sets (like login IDs), it is not a good choice though.

bostik replied:
Hum, true. I'll think about how to express these nicely.

Commit:
Improve protocol documentation, step 1

This step combines a number of clarifications and improvements.

- emphasise on-the-wire encoding requirement
- clarify protocol description
- clarify support for accepted value types
- add a note about timer and histogram equality
- show where alternate uses may be useful

Things left out from this step:
- value ranges with different TSDB implementations
- rate handling

Signed-off-by: Mika Boström <bostik@iki.fi>

jsiembida commented:
LGTM. Can I merge it, or there is more to come?

bostik commented:
I think this is good for the initial release. As you said, the topic of values and their ranges needs to be taken into account separately. So let's make that an edit on its own merit.
